app/controllers/polls_controller.rb

```diff
@@ -67,13 +67,19 @@ def poll_params
   end
 
   def remember_id(id)
-    cookies[:voted_polls] ||= '[]'
-    arr = JSON.parse(cookies[:voted_polls])
+    cookies.signed.permanent[:voted_polls] ||= '[]'
+    arr = JSON.parse(cookies.signed[:voted_polls])
     arr << id
-    cookies[:voted_polls] = JSON.generate(arr.uniq)
+    cookies.signed[:voted_polls] = JSON.generate(arr.uniq)
   end
 
   def remembered_ids
-    JSON.parse(cookies[:voted_polls] || '[]').map(&:to_i)
+    JSON.parse(cookies.signed[:voted_polls] || '[]').map(&:to_i)
+  end
+
+  def save_preferences_as_weight(poll, arr)
+    arr = arr.map { |opt| opt.split('_').last.to_i }.reverse
+    poll.vote_results << poll.options.ids.map { |opt_id| arr.index opt_id }
+    poll.save
```

Comment:
I think you're calling `poll.save` twice when `choose` action is performed.
This can cause some data inconsistency — imagine `poll.save` succeeds here, but `poll.vote!` failes some validation — will DB state be correct?

If saving poll twice cannot be avoided — please consider wrapping the `choose` action code in a transaction.